Allow generic solver to be passed to Newton method

rsenne opened the pull request:
This PR is in reference to issue #1170. To solve this I have edited the interface to allow a user to pass a generic solver. This in theory would allow one to use any matrix type given the user has that package loaded and there exists a solver designed for that matrix type on board. This also allows one to write a custom solver (i.e., problem specific factorization, strategies for solving) to circumvent the default behavior. I have included some examples of this in testing including:

1. Different factorizations (would allow users to circumvent the default PositiveFactorizations.jl implementation of cholesky)
2. Usage of a non-originally supported matrix type (StaticArray)
4. Test type preservation of Hessian with TwiceDifferentiable object using BlockArrays.jl
3. A more realistic example of a non-trivial use case (block-tridiagonal hessian) this implementation is not totally optimized as it is a proof of principle

Let me know if there are any other tests you would like me to include or changes to the documentation.

pkofod commented on src/multivariate/solvers/second_order/newton.jl:
```diff
@@ -1,90 +1,88 @@
+
+The `solve` function should take (H, g) and return s such that H*s = -g.
+Defaults to a robust solver that handles dense or sparse matrices.
+To do this, we use a special factorization from the package `PositiveFactorizations.jl` to ensure
```
```suggestion
If the matrix is not an `AbstractSparseMatrix`, we use a special factorization from the package `PositiveFactorizations.jl` to ensure
```

pkofod commented on src/multivariate/solvers/second_order/newton.jl:
```diff
@@ -1,90 +1,88 @@
+    if H isa AbstractSparseMatrix
+        return -(H \ g)
+    else
+        # Use PositiveFactorizations for robustness on dense matrices
```
```suggestion
        # Use PositiveFactorizations for robustness on dense matrices
         # Search direction is always the negative gradient divided by
         # a matrix encoding the absolute values of the curvatures
         # represented by H. It deviates from the usual "add a scaled
         # identity matrix" version of the modified Newton method. More
         # information can be found in the discussion at issue #153.
```

rsenne commented:
Hi @devmotion. You're right that's my mistake. In honesty I somewhat forgot what I was doing on this PR so my apologies for the late reply. I think i have fixed the regression I introduced. Let me know if there are any other issues

devmotion commented on src/multivariate/solvers/second_order/newton.jl:
```diff
@@ -40,51 +44,55 @@ end
+         # information can be found in the discussion at issue #153.
+         state.F = cholesky!(Positive, H)
+         if g isa Array
+            ldiv!(state.s, state.F, -g)
```
Use
```suggestion
            ldiv!(state.s, state.F, g)
            state.s .= .-state.s
```
to reduce allocations?

rsenne replied:
From the original code it looks like the gradient was negated?

```
ldiv!(state.s, state.F, -gradient(d))
```

Otherwise yes to the suggestion for reduced allocations.

rsenne commented on src/multivariate/solvers/second_order/newton.jl:
```diff
@@ -40,51 +44,55 @@ end
+         # identity matrix" version of the modified Newton method. More
+         # information can be found in the discussion at issue #153.
+         state.F = cholesky!(Positive, H)
+         if g isa Array
```
There was an original comment in the code that said:

```
# is this actually StridedArray?
```
I think this is true, but didn't want to deviate from the source. But i will commit this change.

rsenne commented on src/multivariate/solvers/second_order/newton.jl:
```diff
@@ -40,51 +44,55 @@ end
+            ldiv!(state.s, state.F, -g)
+         else
+            gv = convert(Vector{T}, length(g))
+            copyto!(gv, -g)
```
I agree.
